File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isValidBST(TreeNode root) {
        ArrayList<Integer> arr = new ArrayList<>();
        generate(root, arr);
        int prev = arr.get(0) - 1;
        for (Integer i : arr) {
            if (i <= prev) {
                return false;
            }
            prev = i;
        }
        return true;
    }

    public void generate(TreeNode root, ArrayList<Integer> arr) {
        if (root != null) {
            generate(root.left, arr);
            arr.add(root.val);
            generate(root.right, arr);
        }
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def helper(self, root):
        if not root:
            return 0
        
        l = self.helper(root.left)
        r = self.helper(root.right)
        self.best = max(l + r, self.best)
        '''
                    1
                                2
                        3              4
                            5               6
        '''

        return max(l, r) + 1
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        '''
                    1
            2               2
        3       3       3       3
        '''
        self.best = 0
        l = self.helper(root.left)
        r = self.helper(root.right)
        return max(self.best, l+r)

        

